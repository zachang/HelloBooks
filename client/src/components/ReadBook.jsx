import React, { Component } from 'react';
import PropTypes from 'react-proptypes';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PDFReader from 'react-pdf-reader';
import 'react-pdf-reader/dist/sass/TextLayerBuilder.scss';
import 'react-pdf-reader/dist/sass/PdfReader.scss';
import { getOneBookAction } from '../actions/bookAction';

/**
 * ReadBook class declaration
 *
 * @class ReadBook
 *
 * @extends {React.Component}
 */
class ReadBook extends Component {
  /**
   * class constructor
   *
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      book: null
    };
  }

  /**
   * @method componentWillMount
   *
   * @return {void} void
   */
  componentWillMount() {
    this.props.getOneBookAction(this.props.params.id);
  }

  /**
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps - nextProps
   *
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.book.success === true) {
      this.setState({
        book: nextProps.book.book
      });
    }
  }

  /**
   * @method componentDidMount
   *
   * @return {void} void
   */
  componentDidMount() {
    $('select').material_select();
    $('.collapsible').collapsible();
    $('.dropdown-button').dropdown();
    $('.button-collapse').sideNav();
  }

  /**
   * Renders ReadBook component
   *
   * @return {XML} JSX
   */
  render() {
    const { book } = this.state;

    return (
      <div className='row col m12 col-sm-2'>
        {(book !== null) && <PDFReader
          file={book.bookContent}
          renderType='canvas'
        />
        }
      </div>
    );
  }
}

ReadBook.propTypes = {
  book: PropTypes.object,
  params: PropTypes.object,
  getOneBookAction: PropTypes.func
};
const mapStateToProps = state => ({
  book: state.bookReducer
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ getOneBookAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ReadBook);