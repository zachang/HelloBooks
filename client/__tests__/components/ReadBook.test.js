import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import sinon from 'sinon';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';

import ReadBook from '../../src/components/ReadBook.jsx';


describe('<ReadBook/>', () => {
  // spy on componentWillReceiveProps of ReadBook Page
  sinon.spy(ReadBook.prototype, 'componentWillReceiveProps');

  const props = {
    params: {
      id: 1
    },
    book: {
      success: true,
      book: null
    }
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let mountedComponent;
  beforeEach(() => {
    props.book.success = true;
    store = mockStore(storeConfiguration);
    mountedComponent = mount(<Provider store={store}>
      <ReadBook {...props} />
    </Provider>);
    mountedComponent.setProps({
      book: {
        success: true,
        book: null
      }
    });
  });

  it('renders <ReadBook/> component', () => {
    const tree = toJson(mountedComponent);
    expect(tree).toMatchSnapshot();
  });
});