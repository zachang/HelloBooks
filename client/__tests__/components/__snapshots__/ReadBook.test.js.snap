// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadBook/> renders <ReadBook/> component 1`] = `
<Provider
  book={
    Object {
      "book": null,
      "success": true,
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ReadBook)
    book={
      Object {
        "book": null,
        "success": true,
      }
    }
    params={
      Object {
        "id": 1,
      }
    }
  >
    <ReadBook
      getOneBookAction={[Function]}
      params={
        Object {
          "id": 1,
        }
      }
    >
      <div />
    </ReadBook>
  </Connect(ReadBook)>
</Provider>
`;
